In [1]:
SQL_SOURCE = 'https://biabbiassago.carto.com/api/v2/sql?q='

try:
    #For Python 3
    from urllib.parse import urlparse, urlencode
    from urllib.request import urlopen, Request
    from urllib.error import HTTPError
except ImportError:
    #For Python 2
    from urlparse import urlparse
    from urllib import urlencode
    from urllib2 import urlopen, Request, HTTPError
    
try:
    #for Python 2
    from cStringIO import StringIO as BytesIO
except ImportError:
    #for Python 3
    from io import BytesIO


import ast
import pandas as pd

In [2]:
def queryCartoDB(query, format='CSV', source=SQL_SOURCE):
    '''queries carto datasets from a given carto account
    Arguments: 
    query - string: a valid sql query string
    format - outlut format  OPTIONAL (default CSV)
    source - a valid sql api endpoint OPTIONAL (default carto biabbiassago account)
    Returns:
    the return of the sql query AS A STRING
    NOTES:
    designed for the carto API, tested only with CSV return format'''
    
    data = urlencode({'format': format, 'q': query})
    data=data.encode("utf-8")
    try:
        response = urlopen(source,data)
    except HTTPError as e:
        raise ValueError('\n'.join(ast.literal_eval(e.readline())['error']))
    except Exception:
        raise
    return response.read()

In [3]:
test  = 'SELECT * FROM citibike'

In [4]:
## for visibility purposes only show first 5 entries
pd.read_csv(BytesIO(queryCartoDB(test)), sep=',').head()

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0       NaN  Subscriber                   NaN       1      1965.0   19217   
1       NaN  Subscriber                   NaN       1      1971.0   17575   
2       NaN  Subscriber                   NaN       1      1973.0   20313   
3       NaN  Subscriber                   NaN       2      1978.0   17131   
4       NaN  Subscriber                   NaN       1      1993.0   21289   

   end_station_id  start_station_id  tripduration  cartodb_id  \
0             462               435           351        5960   
1             488               167           643        9806   
2             440               332          1379       13113   
3             423               521           801           1   
4             504               497           379           2   

   end_station_name  start_station_name  end_station_longitude  \
0  W 22 St & 10 Ave     W 21 St & 6 Ave             -74.004519   
1   W 39 St & 9 Ave     E 39 St & 3 Ave             -73.993722   
2   E 45 St & 3 Ave           Cherry St             -73.972826   
3   W 54 St & 9 Ave     8 Ave & W 31 St             -73.986905   
4   1 Ave & E 15 St  E 17 St & Broadway             -73.981656   

   end_station_latitude  start_station_longitude  start_station_latitude  \
0             40.746920               -73.994156               40.741740   
1             40.756458               -73.976049               40.748901   
2             40.752554               -73.979481               40.712199   
3             40.765849               -73.994811               40.750450   
4             40.732219               -73.990093               40.737050   

                 stoptime               starttime  
0  2015-02-01 20:42:00+00  2015-02-01 20:36:00+00  
1  2015-02-03 14:50:00+00  2015-02-03 14:39:00+00  
2  2015-02-04 07:30:00+00  2015-02-04 07:07:00+00  
3  2015-02-01 00:14:00+00  2015-02-01 00:00:00+00  
4  2015-02-01 00:07:00+00  2015-02-01 00:00:00+00

**Extra Credit**
I tested this both in Python 2 and Python 3 Kernels. 
This resource has been very helpful in making the code compatible:
http://python-future.org/compatible_idioms.html#byte-string-literals

## Task 1

**Sort by start_station_id**

In [5]:
query1 = '''
SELECT * FROM citibike 
ORDER BY start_station_id
'''
pd.read_csv(BytesIO(queryCartoDB(query1)), sep=',').head(5)

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0       NaN  Subscriber                   NaN       1      1983.0   21114   
1       NaN  Subscriber                   NaN       1      1964.0   21500   
2       NaN  Subscriber                   NaN       1      1971.0   20139   
3       NaN  Subscriber                   NaN       1      1974.0   17997   
4       NaN    Customer                   NaN       0         NaN   17849   

   end_station_id  start_station_id  tripduration  cartodb_id  \
0             500                72           359       42446   
1              79                72          1678       11298   
2             476                72          1366       19501   
3             352                72           574       20070   
4             328                72          1944        1983   

           end_station_name start_station_name  end_station_longitude  \
0        Broadway & W 51 St   W 52 St & 11 Ave             -73.983362   
1  Franklin St & W Broadway   W 52 St & 11 Ave             -74.006667   
2           E 31 St & 3 Ave   W 52 St & 11 Ave             -73.979661   
3           W 56 St & 6 Ave   W 52 St & 11 Ave             -73.977225   
4   Watts St & Greenwich St   W 52 St & 11 Ave             -74.009660   

   end_station_latitude  start_station_longitude  start_station_latitude  \
0             40.762288               -73.993929               40.767272   
1             40.719116               -73.993929               40.767272   
2             40.743943               -73.993929               40.767272   
3             40.763406               -73.993929               40.767272   
4             40.724055               -73.993929               40.767272   

                 stoptime               starttime  
0  2015-02-07 14:39:00+00  2015-02-07 14:33:00+00  
1  2015-02-03 18:49:00+00  2015-02-03 18:21:00+00  
2  2015-02-04 19:02:00+00  2015-02-04 18:40:00+00  
3  2015-02-04 19:33:00+00  2015-02-04 19:24:00+00  
4  2015-02-01 13:12:00+00  2015-02-01 12:39:00+00

In [6]:
# here we show the end of the dataset... the largest stations. 
pd.read_csv(BytesIO(queryCartoDB(query1)), sep=',').tail(5)

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
46195       NaN  Subscriber                   NaN       1      1979.0   18470   
46196       NaN  Subscriber                   NaN       2      1985.0   16783   
46197       NaN  Subscriber                   NaN       2      1991.0   15456   
46198       NaN  Subscriber                   NaN       1      1954.0   18797   
46199       NaN  Subscriber                   NaN       1      1966.0   20131   

       end_station_id  start_station_id  tripduration  cartodb_id  \
46195             363              3002           108       41340   
46196             195              3002           269       24333   
46197            2008              3002           272       12433   
46198             427              3002           418       24329   
46199             427              3002           545       24326   

            end_station_name          start_station_name  \
46195         West Thames St  South End Ave & Liberty St   
46196  Liberty St & Broadway  South End Ave & Liberty St   
46197  Little West St & 1 Pl  South End Ave & Liberty St   
46198    Bus Slip & State St  South End Ave & Liberty St   
46199    Bus Slip & State St  South End Ave & Liberty St   

       end_station_longitude  end_station_latitude  start_station_longitude  \
46195             -74.017134             40.708347               -74.015756   
46196             -74.010434             40.709056               -74.015756   
46197             -74.016777             40.705693               -74.015756   
46198             -74.013942             40.701907               -74.015756   
46199             -74.013942             40.701907               -74.015756   

       start_station_latitude                stoptime               starttime  
46195               40.711512  2015-02-07 12:09:00+00  2015-02-07 12:07:00+00  
46196               40.711512  2015-02-05 09:31:00+00  2015-02-05 09:27:00+00  
46197               40.711512  2015-02-03 21:57:00+00  2015-02-03 21:53:00+00  
46198               40.711512  2015-02-05 09:34:00+00  2015-02-05 09:27:00+00  
46199               40.711512  2015-02-05 09:36:00+00  2015-02-05 09:27:00+00

**Sort by tripduration**

In [7]:
query2 = '''
SELECT * FROM citibike 
ORDER BY tripduration
'''
pd.read_csv(BytesIO(queryCartoDB(query2)), sep=',').head(5)

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0       NaN  Subscriber                   NaN       1      1983.0   18581   
1       NaN  Subscriber                   NaN       1      1986.0   17643   
2       NaN  Subscriber                   NaN       1      1981.0   21524   
3       NaN  Subscriber                   NaN       1      1990.0   21644   
4       NaN  Subscriber                   NaN       2      1960.0   21627   

   end_station_id  start_station_id  tripduration  cartodb_id  \
0            2009              2009            60       24408   
1             379               379            60        7798   
2             393               393            60       45697   
3             296               296            60       14117   
4             512               512            60       13482   

           end_station_name        start_station_name  end_station_longitude  \
0  Catherine St & Monroe St  Catherine St & Monroe St             -73.996826   
1           W 31 St & 7 Ave           W 31 St & 7 Ave             -73.991600   
2         E 5 St & Avenue C         E 5 St & Avenue C             -73.979955   
3      Division St & Bowery      Division St & Bowery             -73.997047   
4           W 29 St & 9 Ave           W 29 St & 9 Ave             -73.998393   

   end_station_latitude  start_station_longitude  start_station_latitude  \
0             40.711174               -73.996826               40.711174   
1             40.749156               -73.991600               40.749156   
2             40.722992               -73.979955               40.722992   
3             40.714131               -73.997047               40.714131   
4             40.750073               -73.998393               40.750073   

                 stoptime               starttime  
0  2015-02-05 09:33:00+00  2015-02-05 09:32:00+00  
1  2015-02-02 21:40:00+00  2015-02-02 21:39:00+00  
2  2015-02-07 21:41:00+00  2015-02-07 21:40:00+00  
3  2015-02-04 08:42:00+00  2015-02-04 08:41:00+00  
4  2015-02-04 07:53:00+00  2015-02-04 07:52:00+00

In [8]:
pd.read_csv(BytesIO(queryCartoDB(query2)), sep=',').tail(5)

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
46195       NaN  Subscriber                   NaN       1      1977.0   20527   
46196       NaN  Subscriber                   NaN       2      1966.0   17016   
46197       NaN  Subscriber                   NaN       1      1982.0   18425   
46198       NaN  Subscriber                   NaN       2      1951.0   19209   
46199       NaN  Subscriber                   NaN       1      1955.0   20263   

       end_station_id  start_station_id  tripduration  cartodb_id  \
46195             389               438         37665        6670   
46196             450               449         41908       39346   
46197             387               195         42784        7702   
46198             459               492         43013       46083   
46199             250               263         43016        7816   

               end_station_name        start_station_name  \
46195       Broadway & Berry St       St Marks Pl & 1 Ave   
46196           W 49 St & 8 Ave           W 52 St & 9 Ave   
46197   Centre St & Chambers St     Liberty St & Broadway   
46198          W 20 St & 11 Ave           W 33 St & 7 Ave   
46199  Lafayette St & Jersey St  Elizabeth St & Hester St   

       end_station_longitude  end_station_latitude  start_station_longitude  \
46195             -73.965251             40.710446               -73.985649   
46196             -73.987882             40.762272               -73.987895   
46197             -74.004607             40.712733               -74.010434   
46198             -74.007756             40.746745               -73.990931   
46199             -73.995653             40.724561               -73.996375   

       start_station_latitude                stoptime               starttime  
46195               40.727791  2015-02-02 18:57:00+00  2015-02-02 08:29:00+00  
46196               40.764618  2015-02-07 11:09:00+00  2015-02-06 23:31:00+00  
46197               40.709056  2015-02-03 08:00:00+00  2015-02-02 20:07:00+00  
46198               40.750200  2015-02-08 11:20:00+00  2015-02-07 23:23:00+00  
46199               40.717290  2015-02-03 09:57:00+00  2015-02-02 22:00:00+00

**Get top 10 records, i.e. 10 shortest trip records**

In [9]:
query3 = '''
SELECT * FROM citibike
ORDER BY tripduration
LIMIT 10
'''

In [10]:
pd.read_csv(BytesIO(queryCartoDB(query3)), sep=',')

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0       NaN  Subscriber                   NaN       1        1971   14788   
1       NaN  Subscriber                   NaN       1        1990   21644   
2       NaN  Subscriber                   NaN       1        1982   19896   
3       NaN  Subscriber                   NaN       1        1980   16340   
4       NaN  Subscriber                   NaN       1        1985   20572   
5       NaN  Subscriber                   NaN       2        1960   21627   
6       NaN  Subscriber                   NaN       1        1956   20842   
7       NaN  Subscriber                   NaN       1        1986   17643   
8       NaN  Subscriber                   NaN       1        1980   16148   
9       NaN  Subscriber                   NaN       1        1970   20309   

   end_station_id  start_station_id  tripduration  cartodb_id  \
0             521               521            60       13043   
1             296               296            60       14117   
2             251               251            60        8284   
3             471               471            60       11104   
4             317               317            60        3110   
5             512               512            60       13482   
6             474               474            60        7299   
7             379               379            60        7798   
8             523               523            60       21407   
9             248               248            60       19243   

          end_station_name       start_station_name  end_station_longitude  \
0          8 Ave & W 31 St          8 Ave & W 31 St             -73.994811   
1     Division St & Bowery     Division St & Bowery             -73.997047   
2      Mott St & Prince St      Mott St & Prince St             -73.994800   
3  Grand St & Havemeyer St  Grand St & Havemeyer St             -73.956981   
4        E 6 St & Avenue B        E 6 St & Avenue B             -73.981854   
5          W 29 St & 9 Ave          W 29 St & 9 Ave             -73.998393   
6          5 Ave & E 29 St          5 Ave & E 29 St             -73.986831   
7          W 31 St & 7 Ave          W 31 St & 7 Ave             -73.991600   
8          W 38 St & 8 Ave          W 38 St & 8 Ave             -73.991382   
9    Laight St & Hudson St    Laight St & Hudson St             -74.007718   

   end_station_latitude  start_station_longitude  start_station_latitude  \
0             40.750450               -73.994811               40.750450   
1             40.714131               -73.997047               40.714131   
2             40.723180               -73.994800               40.723180   
3             40.712868               -73.956981               40.712868   
4             40.724537               -73.981854               40.724537   
5             40.750073               -73.998393               40.750073   
6             40.745168               -73.986831               40.745168   
7             40.749156               -73.991600               40.749156   
8             40.754666               -73.991382               40.754666   
9             40.721854               -74.007718               40.721854   

                 stoptime               starttime  
0  2015-02-04 06:55:00+00  2015-02-04 06:54:00+00  
1  2015-02-04 08:42:00+00  2015-02-04 08:41:00+00  
2  2015-02-03 08:08:00+00  2015-02-03 08:07:00+00  
3  2015-02-03 18:02:00+00  2015-02-03 18:01:00+00  
4  2015-02-01 14:35:00+00  2015-02-01 14:34:00+00  
5  2015-02-04 07:53:00+00  2015-02-04 07:52:00+00  
6  2015-02-02 17:15:00+00  2015-02-02 17:14:00+00  
7  2015-02-02 21:40:00+00  2015-02-02 21:39:00+00  
8  2015-02-04 22:39:00+00  2015-02-04 22:38:00+00  
9  2015-02-04 18:23:00+00  2015-02-04 18:22:00+00

** Get last 10 records, i.e. 10 longest trip records **

In [11]:
query4 = '''
SELECT * FROM citibike
ORDER BY tripduration DESC
LIMIT 10
'''

In [12]:
pd.read_csv(BytesIO(queryCartoDB(query4)), sep=',')

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0       NaN  Subscriber                   NaN       1      1955.0   20263   
1       NaN  Subscriber                   NaN       2      1951.0   19209   
2       NaN  Subscriber                   NaN       1      1982.0   18425   
3       NaN  Subscriber                   NaN       2      1966.0   17016   
4       NaN  Subscriber                   NaN       1      1977.0   20527   
5       NaN    Customer                   NaN       0         NaN   14691   
6       NaN  Subscriber                   NaN       1      1967.0   18664   
7       NaN    Customer                   NaN       0         NaN   20703   
8       NaN  Subscriber                   NaN       1      1982.0   14769   
9       NaN  Subscriber                   NaN       2      1974.0   16802   

   end_station_id  start_station_id  tripduration  cartodb_id  \
0             250               263         43016        7816   
1             459               492         43013       46083   
2             387               195         42784        7702   
3             450               449         41908       39346   
4             389               438         37665        6670   
5             525               453         35251          78   
6            2009               494         35127        3830   
7             372                83         33647       20353   
8             490               297         32634       12444   
9             488               497         31396        7981   

            end_station_name             start_station_name  \
0   Lafayette St & Jersey St       Elizabeth St & Hester St   
1           W 20 St & 11 Ave                W 33 St & 7 Ave   
2    Centre St & Chambers St          Liberty St & Broadway   
3            W 49 St & 8 Ave                W 52 St & 9 Ave   
4        Broadway & Berry St            St Marks Pl & 1 Ave   
5           W 34 St & 11 Ave                W 22 St & 8 Ave   
6   Catherine St & Monroe St                W 26 St & 8 Ave   
7  Franklin Ave & Myrtle Ave  Atlantic Ave & Fort Greene Pl   
8            8 Ave & W 33 St                E 15 St & 3 Ave   
9            W 39 St & 9 Ave             E 17 St & Broadway   

   end_station_longitude  end_station_latitude  start_station_longitude  \
0             -73.995653             40.724561               -73.996375   
1             -74.007756             40.746745               -73.990931   
2             -74.004607             40.712733               -74.010434   
3             -73.987882             40.762272               -73.987895   
4             -73.965251             40.710446               -73.985649   
5             -74.002116             40.755942               -73.999154   
6             -73.996826             40.711174               -73.997236   
7             -73.958089             40.694528               -73.976323   
8             -73.993934             40.751551               -73.986923   
9             -73.993722             40.756458               -73.990093   

   start_station_latitude                stoptime               starttime  
0               40.717290  2015-02-03 09:57:00+00  2015-02-02 22:00:00+00  
1               40.750200  2015-02-08 11:20:00+00  2015-02-07 23:23:00+00  
2               40.709056  2015-02-03 08:00:00+00  2015-02-02 20:07:00+00  
3               40.764618  2015-02-07 11:09:00+00  2015-02-06 23:31:00+00  
4               40.727791  2015-02-02 18:57:00+00  2015-02-02 08:29:00+00  
5               40.744751  2015-02-01 10:40:00+00  2015-02-01 00:52:00+00  
6               40.747348  2015-02-02 01:32:00+00  2015-02-01 15:47:00+00  
7               40.683826  2015-02-05 05:12:00+00  2015-02-04 19:52:00+00  
8               40.734232  2015-02-04 07:04:00+00  2015-02-03 22:00:00+00  
9               40.737050  2015-02-03 15:03:00+00  2015-02-03 06:19:00+00

In [13]:
# Longest 10 under 3 hours

query5 = '''
SELECT * FROM citibike 
WHERE tripduration<10800
ORDER BY tripduration DESC
LIMIT 10
'''

pd.read_csv(BytesIO(queryCartoDB(query5)), sep=',')

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0       NaN  Subscriber                   NaN       1        1973   15082   
1       NaN  Subscriber                   NaN       1        1950   16099   
2       NaN  Subscriber                   NaN       1        1983   21323   
3       NaN  Subscriber                   NaN       2        1987   19921   
4       NaN  Subscriber                   NaN       1        1970   19070   
5       NaN  Subscriber                   NaN       1        1985   18088   
6       NaN  Subscriber                   NaN       1        1962   14726   
7       NaN  Subscriber                   NaN       2        1966   14691   
8       NaN  Subscriber                   NaN       1        1940   14802   
9       NaN  Subscriber                   NaN       1        1968   15018   

   end_station_id  start_station_id  tripduration  cartodb_id  \
0             173               284         10757       41392   
1             435               512         10660       10912   
2             405               405         10174       40221   
3             264               386          9969        7472   
4            2023               468          9938        9696   
5             435               435          9807       15336   
6            2021               457          9711       25203   
7             517               402          9572       19897   
8             499               499          9556        7073   
9             324               324          9373       11810   

                end_station_name             start_station_name  \
0             Broadway & W 49 St          Greenwich Ave & 8 Ave   
1                W 21 St & 6 Ave                W 29 St & 9 Ave   
2  Washington St & Gansevoort St  Washington St & Gansevoort St   
3           Maiden Ln & Pearl St           Centre St & Worth St   
4        E 55 St & Lexington Ave             Broadway & W 55 St   
5                W 21 St & 6 Ave                W 21 St & 6 Ave   
6                W 45 St & 8 Ave             Broadway & W 58 St   
7          Pershing Square South             Broadway & E 22 St   
8             Broadway & W 60 St             Broadway & W 60 St   
9        DeKalb Ave & Hudson Ave        DeKalb Ave & Hudson Ave   

   end_station_longitude  end_station_latitude  start_station_longitude  \
0             -73.984427             40.760647               -74.002638   
1             -73.994156             40.741740               -73.998393   
2             -74.008119             40.739323               -74.008119   
3             -74.007319             40.707065               -74.002345   
4             -73.970314             40.759681               -73.981923   
5             -73.994156             40.741740               -73.994156   
6             -73.988597             40.759291               -73.981693   
7             -73.977910             40.751581               -73.989551   
8             -73.981918             40.769155               -73.981918   
9             -73.981013             40.689888               -73.981013   

   start_station_latitude                stoptime               starttime  
0               40.739017  2015-02-07 15:16:00+00  2015-02-07 12:17:00+00  
1               40.750073  2015-02-03 20:40:00+00  2015-02-03 17:42:00+00  
2               40.739323  2015-02-07 12:20:00+00  2015-02-07 09:31:00+00  
3               40.714948  2015-02-02 21:08:00+00  2015-02-02 18:22:00+00  
4               40.765265  2015-02-03 16:55:00+00  2015-02-03 14:10:00+00  
5               40.741740  2015-02-04 13:36:00+00  2015-02-04 10:53:00+00  
6               40.766953  2015-02-05 13:54:00+00  2015-02-05 11:12:00+00  
7               40.740343  2015-02-04 21:48:00+00  2015-02-04 19:09:00+00  
8               40.769155  2015-02-02 16:05:00+00  2015-02-02 13:25:00+00  
9               40.689888  2015-02-03 22:01:00+00  2015-02-03 19:25:00+00

## TASK 2

** All trip records that started on Feb-02-2015**

In [14]:
query6 = '''
SELECT * FROM citibike
WHERE starttime >= '2015-02-02 00:00'
AND starttime < '2015-02-03 00:00'
'''
pd.read_csv(BytesIO(queryCartoDB(query6)), sep=',').head()

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0       NaN  Subscriber                   NaN       1      1992.0   20684   
1       NaN  Subscriber                   NaN       2      1964.0   16094   
2       NaN  Subscriber                   NaN       1      1974.0   17362   
3       NaN  Subscriber                   NaN       2      1992.0   15475   
4       NaN  Subscriber                   NaN       1      1968.0   16272   

   end_station_id  start_station_id  tripduration  cartodb_id  \
0             489               442           199        6442   
1             349               326           418        6443   
2             415               279           276        6444   
3             237               496           420        6445   
4             450               477           304        6446   

            end_station_name    start_station_name  end_station_longitude  \
0           10 Ave & W 28 St       W 27 St & 7 Ave             -74.001768   
1    Rivington St & Ridge St       E 11 St & 1 Ave             -73.983299   
2  Pearl St & Hanover Square  Peck Slip & Front St             -74.009260   
3            E 11 St & 2 Ave       E 16 St & 5 Ave             -73.986724   
4            W 49 St & 8 Ave       W 41 St & 8 Ave             -73.987882   

   end_station_latitude  start_station_longitude  start_station_latitude  \
0             40.750664               -73.993915               40.746647   
1             40.718502               -73.984267               40.729538   
2             40.704718               -74.001670               40.707873   
3             40.730473               -73.992390               40.737262   
4             40.762272               -73.990026               40.756405   

                 stoptime               starttime  
0  2015-02-02 00:05:00+00  2015-02-02 00:02:00+00  
1  2015-02-02 00:10:00+00  2015-02-02 00:03:00+00  
2  2015-02-02 00:09:00+00  2015-02-02 00:04:00+00  
3  2015-02-02 00:13:00+00  2015-02-02 00:06:00+00  
4  2015-02-02 00:12:00+00  2015-02-02 00:07:00+00

** Retrieve all trip records started on a weekend**

In [15]:
query7 = '''
SELECT * FROM citibike
WHERE extract(DOW FROM starttime) IN (0,6)
'''

# head just for visibility purposes...
pd.read_csv(BytesIO(queryCartoDB(query7)), sep=',').head()

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0       NaN  Subscriber                   NaN       1      1965.0   19217   
1       NaN  Subscriber                   NaN       2      1978.0   17131   
2       NaN  Subscriber                   NaN       1      1993.0   21289   
3       NaN  Subscriber                   NaN       2      1969.0   18903   
4       NaN  Subscriber                   NaN       2      1985.0   21044   

   end_station_id  start_station_id  tripduration  cartodb_id  \
0             462               435           351        5960   
1             423               521           801           1   
2             504               497           379           2   
3             127               281          2474           3   
4             505              2004           818           4   

        end_station_name                 start_station_name  \
0       W 22 St & 10 Ave                    W 21 St & 6 Ave   
1        W 54 St & 9 Ave                    8 Ave & W 31 St   
2        1 Ave & E 15 St                 E 17 St & Broadway   
3  Barrow St & Hudson St  Grand Army Plaza & Central Park S   
4        6 Ave & W 33 St                  6 Ave & Broome St   

   end_station_longitude  end_station_latitude  start_station_longitude  \
0             -74.004519             40.746920               -73.994156   
1             -73.986905             40.765849               -73.994811   
2             -73.981656             40.732219               -73.990093   
3             -74.006744             40.731724               -73.973715   
4             -73.988484             40.749013               -74.004704   

   start_station_latitude                stoptime               starttime  
0               40.741740  2015-02-01 20:42:00+00  2015-02-01 20:36:00+00  
1               40.750450  2015-02-01 00:14:00+00  2015-02-01 00:00:00+00  
2               40.737050  2015-02-01 00:07:00+00  2015-02-01 00:00:00+00  
3               40.764397  2015-02-01 00:42:00+00  2015-02-01 00:01:00+00  
4               40.724399  2015-02-01 00:15:00+00  2015-02-01 00:01:00+00

** FIND average count min and max for weekend trips** 

In [16]:
query8 = '''
SELECT
COUNT(tripduration) as count,
AVG(tripduration) as avg,
MIN(tripduration) as min,
MAX(tripduration) as max

FROM citibike

WHERE extract(DOW FROM starttime) IN (0,6)
'''
pd.read_csv(BytesIO(queryCartoDB(query8)), sep=',')

count         avg  min    max
0  13231  662.942181   60  43013

** All trip records started on weekday **

In [17]:
query9 = '''
SELECT * FROM citibike
WHERE extract(DOW FROM starttime) NOT IN (0,6)
'''
pd.read_csv(BytesIO(queryCartoDB(query9)), sep=',').head()

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0       NaN  Subscriber                   NaN       1      1971.0   17575   
1       NaN  Subscriber                   NaN       1      1973.0   20313   
2       NaN  Subscriber                   NaN       1      1992.0   20684   
3       NaN  Subscriber                   NaN       2      1964.0   16094   
4       NaN  Subscriber                   NaN       1      1974.0   17362   

   end_station_id  start_station_id  tripduration  cartodb_id  \
0             488               167           643        9806   
1             440               332          1379       13113   
2             489               442           199        6442   
3             349               326           418        6443   
4             415               279           276        6444   

            end_station_name    start_station_name  end_station_longitude  \
0            W 39 St & 9 Ave       E 39 St & 3 Ave             -73.993722   
1            E 45 St & 3 Ave             Cherry St             -73.972826   
2           10 Ave & W 28 St       W 27 St & 7 Ave             -74.001768   
3    Rivington St & Ridge St       E 11 St & 1 Ave             -73.983299   
4  Pearl St & Hanover Square  Peck Slip & Front St             -74.009260   

   end_station_latitude  start_station_longitude  start_station_latitude  \
0             40.756458               -73.976049               40.748901   
1             40.752554               -73.979481               40.712199   
2             40.750664               -73.993915               40.746647   
3             40.718502               -73.984267               40.729538   
4             40.704718               -74.001670               40.707873   

                 stoptime               starttime  
0  2015-02-03 14:50:00+00  2015-02-03 14:39:00+00  
1  2015-02-04 07:30:00+00  2015-02-04 07:07:00+00  
2  2015-02-02 00:05:00+00  2015-02-02 00:02:00+00  
3  2015-02-02 00:10:00+00  2015-02-02 00:03:00+00  
4  2015-02-02 00:09:00+00  2015-02-02 00:04:00+00

In [18]:
query10 = '''
SELECT
COUNT(tripduration) as count,
AVG(tripduration) as avg,
MIN(tripduration) as min,
MAX(tripduration) as max

FROM citibike

WHERE extract(DOW FROM starttime) NOT IN (0,6)
'''
pd.read_csv(BytesIO(queryCartoDB(query10)), sep=',')

count         avg  min    max
0  32969  681.052292   60  43016

## Task 3 

** Showing list of start locations and number of trips from that station **

In [19]:
query11 = '''
SELECT
    start_station_id,
    COUNT(start_station_id) as numtrips
FROM citibike

GROUP BY
    start_station_id
'''
# all stations and number of trips originating from there
pd.read_csv(BytesIO(queryCartoDB(query11)), sep=',')

start_station_id  numtrips
0                 195       137
1                 251       194
2                 285       373
3                 120        17
4                 483       229
5                 264       115
6                 497       455
7                 305       213
8                 318       486
9                 312       207
10                496       250
11                455       110
12                470       173
13                276        65
14               2008        86
15                313        50
16                262        37
17                362        98
18                151       288
19                384        55
20                119        19
21                399        32
22                356        98
23                270        92
24                437        18
25                514        91
26                518       282
27                329       145
28                268       124
29                531       212
..                ...       ...
297               420        25
298               443        13
299               490       387
300               492       309
301               416        36
302               331        52
303               232        22
304               427        42
305               377       101
306               445       296
307               411        86
308               260        36
309               403       151
310               494       237
311               244        35
312               128       247
313               223       179
314               456       110
315               300       161
316               217        14
317               538       153
318               383       142
319               212       186
320               353        29
321               328        73
322               216        12
323               263       217
324               266        38
325               144         7
326               453       218

[327 rows x 2 columns]

**Get list of start station location (lat, long) with web mercator**

In [20]:
query12 = '''
SELECT
    start_station_id,
    CDB_TransformToWebmercator(CDB_LatLng(
        start_station_latitude,
        start_station_longitude
        )
) as the_geom_webmercator, MIN(cartodb_id) as cartodb_id, start_station_latitude, start_station_longitude
FROM citibike
GROUP BY
      start_station_id,
      start_station_latitude,
      start_station_longitude
'''
pd.read_csv(BytesIO(queryCartoDB(query12)), sep=',').head()

start_station_id                               the_geom_webmercator  \
0               438  0101000020110F0000B77CA834036B5FC19EF5005EBCF7...   
1               342  0101000020110F0000BEA7E6966A6A5FC1E90E1AC83EF6...   
2               242  0101000020110F0000E097312EB1695FC111C4474A72F3...   
3               161  0101000020110F00009B8267C45D6C5FC1577C6302EFF7...   
4               373  0101000020110F0000C12D0F658D675FC1323CC0ADCAF2...   

   cartodb_id  start_station_latitude  start_station_longitude  
0          79               40.727791               -73.985649  
1         609               40.717400               -73.980166  
2         454               40.697883               -73.973503  
3         185               40.729170               -73.998102  
4           6               40.693317               -73.953820

** 500 m from Times Square**

In [21]:
query13 = '''
SELECT
    start_station_id,COUNT(tripduration) as numtrips
FROM citibike
WHERE
     ST_DWithin(CDB_LatLng(
         start_station_latitude,
         start_station_longitude
)::geography, CDB_LatLng(40.7577,-73.9857)::geography, 500)
GROUP BY
      start_station_id,
      start_station_latitude,
      start_station_longitude
'''

pd.read_csv(BytesIO(queryCartoDB(query13)), sep=',')

start_station_id  numtrips
0               477       507
1              2021       141
2               465       251
3               493       141
4               524       112
5               529       221
6               173       213

## Task 4

** Station with longest Average trip duration during weekends 500 m from Times Square**

In [22]:
query14 = '''
SELECT
    start_station_id,
    AVG(tripduration) as avg_duration

FROM citibike
WHERE
     ST_DWithin(
         CDB_LatLng(
              start_station_latitude,
         start_station_longitude
)::geography, CDB_LatLng(40.7577,-73.9857)::geography, 500)

AND extract(DOW FROM starttime) IN (0,6)
GROUP BY
      start_station_id,
      start_station_latitude,
      start_station_longitude
ORDER BY avg_duration DESC
LIMIT 1
'''
pd.read_csv(BytesIO(queryCartoDB(query14)), sep=',')

start_station_id  avg_duration
0               173   1010.104167

The station with the longest average duration within a 500 m radius from Times Square is  is station 173. At this station the average trip duration is of 1010 seconds, i.e. around 17 minutes. 

** TASK 4 EXTRA : lines for visualizaiton in CartoDB**
Create lines for trips started from station within 500m from Times square and lsted less than 2 hours. The number of trips per each pair of stations are output as attrivutes of these lines. 

In [23]:
query15 = '''
SELECT
    start_station_id,
    end_station_id,
    ST_Makeline(
         CDB_TransformToWebmercator(CDB_LatLng(
              start_station_latitude,
              start_station_longitude
         )),
         CDB_TransformToWebmercator(CDB_LatLng(
              end_station_latitude,
              end_station_longitude))
    ) as the_geom_webmercator, 
    MIN(cartodb_id) as cartodb_id,
    AVG(tripduration) as avg_duration

FROM citibike

WHERE
    tripduration<7200
and
     ST_DWithin(
         CDB_LatLng(
              start_station_latitude,
                start_station_longitude
            )::geography,
            CDB_LatLng(40.7577,-73.9857)::geography, 500)

GROUP BY
    start_station_id,
    start_station_latitude,
    start_station_longitude,
    end_station_id,
    end_station_latitude,
    end_station_longitude
    
'''

pd.read_csv(BytesIO(queryCartoDB(query15)), sep=',').head()

start_station_id  end_station_id  \
0               173             528   
1               477             435   
2               465             538   
3               493             512   
4               524              72   

                                the_geom_webmercator  cartodb_id  avg_duration  
0  0102000020110F00000200000020D6742CE16A5FC112F4...       18466         693.0  
1  0102000020110F000002000000F55695027D6B5FC186E5...        2226         895.5  
2  0102000020110F00000200000004AF971C1D6B5FC17A48...       15621         414.5  
3  0102000020110F0000020000009BB87D02B76A5FC146C1...       34734         700.0  
4  0102000020110F000002000000F22A632FBE6A5FC1363A...       37006        1086.0

## ASH - Score 100. Extra Credit 100